datacompy/fuguecompare.py

```diff
@@ -0,0 +1,189 @@
+    abs_tol: float = 0,
+    rel_tol: float = 0,
+    df1_name: str = "df1",
+    df2_name: str = "df2",
+    ignore_spaces: bool = False,
+    ignore_case: bool = False,
+    cast_column_names_lower: bool = True,
+    parallelism: Optional[int] = None,
+) -> bool:
+    """Check whether two dataframes match.
+
+    Both df1 and df2 should be dataframes containing all of the join_columns,
+    with unique column names. Differences between values are compared to
+    abs_tol + rel_tol * abs(df2['value']).
+
+    Parameters
+    ----------
+    df1 : ``AnyDataFrame``
+        First dataframe to check
+    df2 : ``AnyDataFrame``
+        Second dataframe to check
+    join_columns : list or str, optional
+        Column(s) to join dataframes on.  If a string is passed in, that one
+        column will be used.
+    abs_tol : float, optional
+        Absolute tolerance between two values.
+    rel_tol : float, optional
+        Relative tolerance between two values.
+    df1_name : str, optional
+        A string name for the first dataframe.  This allows the reporting to
+        print out an actual name instead of "df1", and allows human users to
+        more easily track the dataframes.
+    df2_name : str, optional
+        A string name for the second dataframe
+    ignore_spaces : bool, optional
+        Flag to strip whitespace (including newlines) from string columns (including any join
+        columns)
+    ignore_case : bool, optional
+        Flag to ignore the case of string columns
+    cast_column_names_lower: bool, optional
+        Boolean indicator that controls of column names will be cast into lower case
+    """
+    if (
+        isinstance(df1, pd.DataFrame)
+        and isinstance(df2, pd.DataFrame)
+        and parallelism is None  # user did not specify parallelism
+        and fa.get_current_parallelism() == 1  # currently on a local execution engine
+    ):
+        comp = Compare(
+            df1=df1,
+            df2=df2,
+            join_columns=join_columns,
+            abs_tol=abs_tol,
+            rel_tol=rel_tol,
+            df1_name=df1_name,
+            df2_name=df2_name,
+            ignore_spaces=ignore_spaces,
+            ignore_case=ignore_case,
+            cast_column_names_lower=cast_column_names_lower,
+        )
+        return comp.matches()
+
+    tdf1 = fa.as_fugue_df(df1)
+    tdf2 = fa.as_fugue_df(df2)
+
+    if isinstance(join_columns, str):
+        hash_cols = [join_columns]
+    else:
+        hash_cols = join_columns
+    # ensure all cols exist
+    hash_cols = tdf1.schema.extract(hash_cols).names
+
+    if cast_column_names_lower:
+        tdf1 = tdf1.rename(
+            {col: col.lower() for col in tdf1.schema.names if col != col.lower()}
+        )
+        tdf2 = tdf2.rename(
+            {col: col.lower() for col in tdf2.schema.names if col != col.lower()}
+        )
+        hash_cols = [col.lower() for col in hash_cols]
+
+    if tdf1.schema != tdf2.schema:
```

Comment:
I think this can be removed I'm not mistaken. The internals of Compare will try casting and various check to see if will work.

Reply:
This is to fail fast, it can fail without any real compute. For big data this can save a lot of time and money. But yeah, shall we make this optional?

Reply:
:+1:  I think the optional makes sense. Something like `strict_schema`?

But in the case where the "types" are different (but names match) then I think we should check to make sure we are checking. Like how the pandas compare behaves.

Reply:
Maybe something along the lines of: 

```python
    if strict_schema:
        if tdf1.schema != tdf2.schema:
            return False

    # check that hash columns exist
    if hash_cols not in tdf1.schema or hash_cols not in tdf2.schema:
        return False
```

Reply:
Yes I agree with strict schema. But for missing cols, I ensured that here:
```
hash_cols = tdf1.schema.extract(hash_cols).names
```
If the join cols user provided are not found in the schema, I think it should raise an exception. This is not data issue, the user breaks the contract.

Reply:
Agreed! We would also need to check `tdf2` to ensure it doesn't throw an exception here too.

Reply:
So should I make the change, or you can cover that?